src/backend/drm/gbm/surface.rs

```diff
@@ -1,97 +1,146 @@
+                warn!(self.logger, "Tried to swap with an already queued flip");
+                return Err(Error::FrontBuffersExhausted);
+            }
 
-        if cfg!(debug_assertions) {
-            if let Err(err) = self.crtc.get_framebuffer(fb) {
-                error!(self.logger, "Cached framebuffer invalid: {:?}: {}", fb, err);
+            // supporting only one buffer would cause a lot of inconvinience and
+            // would most likely result in a lot of flickering.
+            // neither weston, wlc or wlroots bother with that as well.
+            // so we just assume we got at least two buffers to do flipping.
+            let mut next_bo = self
+                .surface
+                .lock().unwrap()
+                .lock_front_buffer()
+                .map_err(|_| Error::FrontBufferLockFailed)?;
+
+            // create a framebuffer if the front buffer does not have one already
+            // (they are reused by gbm)
+            let maybe_fb = next_bo
+                .userdata()
+                .map_err(|_| Error::InvalidInternalState)?
+                .cloned();
+            let fb = if let Some(info) = maybe_fb {
+                info
+            } else {
+                let fb = self
+                    .crtc
+                    .add_planar_framebuffer(&next_bo, &[0; 4], 0)
+                    .compat()
+                    .map_err(Error::FramebufferCreationFailed)?;
+                next_bo.set_userdata(fb).unwrap();
+                fb
+            };
+            buffers.next_buffer = Some(next_bo);
+
+            if cfg!(debug_assertions) {
+                if let Err(err) = self.crtc.get_framebuffer(fb) {
+                    error!(self.logger, "Cached framebuffer invalid: {:?}: {}", fb, err);
+                }
             }
-        }
 
-        let result = if self.recreated.get() {
-            debug!(self.logger, "Commiting new state");
-            self.crtc.commit(fb).map_err(Error::Underlying)
-        } else {
-            trace!(self.logger, "Queueing Page flip");
-            RawSurface::page_flip(&self.crtc, fb).map_err(Error::Underlying)
+            (
+                if self.recreated.load(Ordering::SeqCst) {
+                    debug!(self.logger, "Commiting new state");
+                    self.crtc.commit(fb).map_err(Error::Underlying)
+                } else {
+                    trace!(self.logger, "Queueing Page flip");
+                    RawSurface::page_flip(&self.crtc, fb).map_err(Error::Underlying)
+                },
+                fb
+            )
         };
 
         match result {
             Ok(_) => {
-                self.recreated.set(false);
-                self.current_frame_buffer.set(Some(fb));
+                self.recreated.store(false, Ordering::SeqCst);
+                let mut buffers = self.buffers.lock().unwrap();
+                buffers.current_frame_buffer = Some(fb);
                 Ok(())
             }
             Err(err) => {
```

Comment:
`unlock_buffer` also locks `.buffers`, so we need to make sure that the scope above is closed already, before entering this function, otherwise we deadlock. (std::sync::Mutex is not re-entrant safe.)